except_star.md

```diff
@@ -234,20 +284,177 @@ except *ValueError as errors:
 # where the `RuntimeError()` instance would have
 # its __cause__ attribute set to
 #
-#   ExceptionGroup({
-#     ValueError('a'), ValueError('b')
-#   })
+#   ExceptionGroup(
+#     ValueError('a'),
+#     ValueError('b')
+#   )
+```
+
+### Recursive Matching
+
+The matching of `except *` clauses against an `ExceptionGroup` is performed
+recursively. E.g.:
+
+```python
+try:
+  raise ExceptionGroup(
+    ValueError('a'),
+    TypeError('b'),
+    ExceptionGroup(
+      TypeError('c'),
+      KeyError('d')
+    )
+  )
+except *TypeError as e:
+  print(f'got some TypeErrors: {len(e)}')
+except *Exception:
+  pass
+
+# would print:
+#
+#  got some TypeErrors: 2
+```
+
+Iteration over an `ExceptionGroup` that has nested `ExceptionGroup` objects
+in it effectively flattens the entire tree. E.g.
+
+```python
+print(
+  list(
+    ExceptionGroup(
+      ValueError('a'),
+      TypeError('b'),
+      ExceptionGroup(
+        TypeError('c'),
+        KeyError('d')
+      )
+    )
+  )
+)
+
+# would output:
+#
+#   [ValueError('a'), TypeError('b'), TypeError('c'), KeyError('d')]
+```
+
+### "raise e" vs "raise"
+
+The difference between bare `raise` and a more specific `raise e` is more
+significant for exception groups than it is for regular exceptions. Consider
+the following two examples that illustrate it.
+
+Bare `raise` preserves the exception group internal structure:
+
+```python
+try:
+  raise ExceptionGroup(
+    ValueError('a'),
+    TypeError('b'),
+    ExceptionGroup(
+      TypeError('c'),
+      KeyError('d')
+    )
+  )
+except *TypeError as e:
+  raise
+
+# would crash with:
+#
+#  ExceptionGroup(
+#    ValueError('a'),
+#    TypeError('b'),
+#    ExceptionGroup(
+#      TypeError('c'),
+#      KeyError('d')
+#    )
+#  )
+```
+
+Whereas `raise e` would flatten the captured subset:
+
+```python
+try:
+  raise ExceptionGroup(
+    ValueError('a'),
+    TypeError('b'),
+    ExceptionGroup(
+      TypeError('c'),
+      KeyError('d')
+    )
+  )
+except *TypeError as e:
+  raise e
+
+# would crash with:
+#
+#  ExceptionGroup(
+#    ValueError('a'),
+#    ExceptionGroup(
+#      TypeError('b'),
+#      TypeError('c'),
+#    ),
+#    ExceptionGroup(
+#      KeyError('d')
+#    )
+#  )
```

Comment:
Yikes! But why would it keep that group? If you start with `ExceptionGroup(A(), B())` and you write
```
except *B as e:
    raise e
```
I hope you wouldn't end up with `ExceptionGroup(A(), ExceptionGroup(B()))`?

Reply:
I hoped you would comment on this! I feel like this is one of the trickiest behaviors for us to define.

> If you start with ExceptionGroup(A(), B()) and you write
> ```
> except *B as e:
>     raise e
> ```
> I hope you wouldn't end up with ExceptionGroup(A(), ExceptionGroup(B()))?

I think it should end up with `ExceptionGroup(A(), ExceptionGroup(B()))`. What if we change your code to:

```
except *B as e:
    e.marker = 'spam'
    raise e
```

I would expect the `marker='spam'` attribute to be present somewhere in what propagates out of the `try` block. And if we merge `e` back and raise `ExceptionGroup(A(), B())` in the end this attribute will be lost.

What do you think?

Reply:
Hm... But should you be doing that? In a sense ‘e’ is ephemeral, only for use in this one block. Surely if you mutate ‘e’ and then use bare raise it should get lost? (Hm, how can we describe bare raise? Ideally it should push the exceptions extracted back exactly from where they were extracted. Maybe the EG API needs to be described more carefully.)

Reply:
> Surely if you mutate ‘e’ and then use bare raise it should get lost?

That's a good observation. So you propose that `raise e` would be equivalent to `raise`. If the user wants to "flatten" the hierarchy that would *explicitly* say `raise ExceptionGroup(*e)`.

I like this idea because it makes it harder for users to accidentally flatten parts of their exceptions tree. The current wording was motivated by the `e.marker = 'spam'` example, but if we treat `e` as an ephemeral object we can perhaps ignore this use case. 

This is also an area where a linter (or mypy, easily) can warn the user that they're working with an ephemeral object and that their changes or references to it will not persist the way they expect them to.

Reply:
I've pushed another update erasing any semantic difference between `raise` and `raise e`.

Reply:
Hm, but there should be a difference, Bare raise doesn’t set ‘—context—‘, but raise e does, and that must be reflected in its inner exceptions.

Reply:
Also ‘raise e from c’ sets ‘—cause—‘ which should broadcast to the inner exceptions.

Reply:
Hm, you mean that bare raise doesn't update the traceback, whereas `raise e` would add that very "raise e" line to the traceback, right? Or is there something about `e.__context__`?

Reply:
Both.

Reply:
Guido, can you explain re `__context__`? I don't see any difference between `raise` and `raise e` here:

```python
def foo():
    try:
        1 / 0
    except ZeroDivisionError:
        try:
            raise ValueError
        except Exception as e:
            raise e

try:
    foo()
except Exception as e:
    print(type(e), type(e.__context__))

# always prints "<class 'ValueError'> <class 'ZeroDivisionError'>"
```

or here:

```python
def foo():
    try:
        1 / 0
    except ZeroDivisionError as e:
        raise e

try:
    foo()
except Exception as e:
    print(type(e), type(e.__context__))

# always prints "<class 'ZeroDivisionError'> <class 'NoneType'>"
```

Reply:
It's not `__context__` but the traceback that's different. Compare:
```
(v39) ~$ cat -n a.py; python a.py 
     1	def foo():
     2	    try:
     3	        1/0
     4	    except ZeroDivisionError as e:
     5	        raise e
     6	
     7	foo()
Traceback (most recent call last):
  File "/Users/guido/a.py", line 7, in <module>
    foo()
  File "/Users/guido/a.py", line 5, in foo
    raise e
  File "/Users/guido/a.py", line 3, in foo
    1/0
ZeroDivisionError: division by zero
```
to
```
(v39) ~$ cat -n b.py; python b.py 
     1	def foo():
     2	    try:
     3	        1/0
     4	    except ZeroDivisionError as e:
     5	        raise
     6	
     7	foo()
Traceback (most recent call last):
  File "/Users/guido/b.py", line 7, in <module>
    foo()
  File "/Users/guido/b.py", line 3, in foo
    1/0
ZeroDivisionError: division by zero
```

Reply:
Got it, i'll push a commit clarifying this soon.